ratelimit/cluster.go

```diff
@@ -1,141 +1,28 @@
-	Values(string) map[string]interface{}
-}
-
-// clusterLimit stores all data required for the cluster ratelimit.
-type clusterLimit struct {
-	group   string
-	local   limiter
-	maxHits int
-	window  time.Duration
-	swarm   Swarmer
-	resize  chan resizeLimit
-	quit    chan struct{}
-}
-
-type resizeLimit struct {
-	s string
-	n int
-}
-
-// newClusterRateLimiter creates a new clusterLimit for given Settings
-// and use the given Swarmer. Group is used in log messages to identify
-// the ratelimit instance and has to be the same in all skipper instances.
-func newClusterRateLimiter(s Settings, sw Swarmer, group string) *clusterLimit {
-	rl := &clusterLimit{
-		group:   group,
-		swarm:   sw,
-		maxHits: s.MaxHits,
-		window:  s.TimeWindow,
-		resize:  make(chan resizeLimit),
-		quit:    make(chan struct{}),
-	}
-	switch s.Type {
-	case ClusterServiceRatelimit:
-		log.Infof("new backend clusterRateLimiter")
-		rl.local = circularbuffer.NewRateLimiter(s.MaxHits, s.TimeWindow)
-	case ClusterClientRatelimit:
-		log.Infof("new client clusterRateLimiter")
-		rl.local = circularbuffer.NewClientRateLimiter(s.MaxHits, s.TimeWindow, s.CleanInterval)
-	default:
-		log.Errorf("Unknown ratelimit type: %s", s.Type)
-		return nil
-	}
-
-	// TODO(sszuecs): we might want to have one goroutine for all of these
-	go func() {
-		for {
-			select {
-			case size := <-rl.resize:
-				log.Debugf("%s resize clusterRatelimit: %v", group, size)
-				// TODO(sszuecs): call with "go" ?
-				rl.Resize(size.s, rl.maxHits/size.n)
-			case <-rl.quit:
-				log.Debugf("%s: quit clusterRatelimit", group)
-				close(rl.resize)
-				return
-			}
-		}
-	}()
-
-	return rl
-}
-
 const swarmPrefix string = `ratelimit.`
 
-// Allow returns true if the request calculated across the cluster of
-// skippers should be allowed else false. It will share it's own data
-// and use the current cluster information to calculate global rates
-// to decide to allow or not.
-func (c *clusterLimit) Allow(s string) bool {
-	key := swarmPrefix + c.group + "." + s
-
-	// t0 is the oldest entry in the local circularbuffer
-	// [ t3, t4, t0, t1, t2]
-	//           ^- current pointer to oldest
-	// now - t0
-	t0 := c.Oldest(s).UTC().UnixNano()
-
-	_ = c.local.Allow(s) // update local rate limit
-
-	if err := c.swarm.ShareValue(key, t0); err != nil {
-		log.Errorf("%s clusterRatelimit failed to share value: %v", c.group, err)
+// newClusterRateLimiter will return a limiter instance, that has a
+// cluster wide knowledge of ongoing requests. Settings are the normal
+// ratelimit settings, Swarmer is an instance sattisfying the Swarmer
+// interface, which is one of swarm.Swarm or noopSwarmer,
+// swarm.Options to configure a swarm.Swarm, RedisOptions to configure
+// redis.Ring and group is the ratelimit group that can span one or
+// multiple routes.
+func newClusterRateLimiter(s Settings, sw Swarmer, so *swarm.Options, ring *ring, group string) limiter {
+	if so != nil {
```

Comment:
we are checking here whether so is nil, but we only use sw (the initialized instance that already consumed the options). We should remove the so *swarm.Options argument.